docs/upgrade.html

```diff
@@ -20,6 +20,45 @@
 <script id="upgrade-template" type="text/x-handlebars-template">
 
 <h4><a id="upgrade_2_1_0" href="#upgrade_2_1_0">Upgrading from 0.8.x, 0.9.x, 0.10.0.x, 0.10.1.x, 0.10.2.x, 0.11.0.x, 1.0.x, 1.1.x, or 2.0.0 to 2.1.0</a></h4>
+
+<p><b>Note that 2.1.x contains an incompatible change to the internal schema used to store consumer offsets. Once the upgrade is
+  complete, it will not be possible to downgrade to previous versions. See the rolling upgrade notes below for more detail.</b></p>
+
+<p><b>For a rolling upgrade:</b></p>
+
+<ol>
+    <li> Update server.properties on all brokers and add the following properties. CURRENT_KAFKA_VERSION refers to the version you
+        are upgrading from. CURRENT_MESSAGE_FORMAT_VERSION refers to the message format version currently in use. If you have previously
+        overridden the message format version, you should keep its current value. Alternatively, if you are upgrading from a version prior
+        to 0.11.0.x, then CURRENT_MESSAGE_FORMAT_VERSION should be set to match CURRENT_KAFKA_VERSION.
+        <ul>
+            <li>inter.broker.protocol.version=CURRENT_KAFKA_VERSION (e.g. 0.8.2, 0.9.0, 0.10.0, 0.10.1, 0.10.2, 0.11.0, 1.0, 1.1).</li>
+            <li>log.message.format.version=CURRENT_MESSAGE_FORMAT_VERSION  (See <a href="#upgrade_10_performance_impact">potential performance impact
+                following the upgrade</a> for the details on what this configuration does.)</li>
+        </ul>
+        If you are upgrading from 0.11.0.x, 1.0.x, 1.1.x, or 2.0.x and you have not overridden the message format, then you only need to override
+        the inter-broker protocol version.
+        <ul>
+            <li>inter.broker.protocol.version=CURRENT_KAFKA_VERSION (0.11.0, 1.0, 1.1, 2.0).</li>
+        </ul>
+    </li>
+    <li> Upgrade the brokers one at a time: shut down the broker, update the code, and restart it. Once you have done so, the
+        brokers will be running the latest version and you can verify that the cluster's behavior and performance meets expectations.
+        It is still possible to downgrade at this point if there are any problems. 
+    </li>
+    <li> Once the entire cluster is upgraded, bump the protocol version by editing <code>inter.broker.protocol.version</code>
+        and setting it to 2.1.
+    </li>
+    <li> Restart the brokers one by one for the new protocol version to take effect. Once the brokers begin using the latest
+        protocol version, it will no longer be possible to downgrade the cluster to an older version.
+    </li>
+    <li> If you have overridden the message format version as instructed above, then you need to do one more rolling restart to
+        upgrade it to its latest version. Once all (or most) consumers have been upgraded to 0.11.0 or later,
```

Comment:
Regarding `Once all (or most) consumers...`, should user also upgrade most producers to 0.11.0 or later before bumping up the message format version? Otherwise, if producer earlier than 0.11.0 produces to Kafka, Kafka server will pay the performance cost of up-conversion, right?

Reply:
This snippet is copied from previous releases. I think the down-conversion costs are perceived to be more significant largely because of consumer fanout. I can add a note about up-conversion as well if you think it is helpful.

Reply:
At LinkedIn the fan-out rate is about 3, which does not seem significantly lager. Where there be considerable (worrisome) performance cost if broker has to down-covert message for every ProduceRequest? If so, it seems simpler and more accurate to replace `Once all (or most) consumers have been upgraded to 0.11.0 or later` with `Once all (or most) producers and consumers have been upgraded to 0.11.0 or later`, and replace `performance cost of down-conversion` with `performance cost of message conversion`.

Reply:
Sure, I've no problem generalizing the comment. I think the historical focus on consumers is likely because of significantly increased memory utilization for down-conversion. This is less of an issue these days because of lazy down-conversion, so that leaves CPU as the primary concern, which can be more evenly spread between producers and consumers depending on the fanout.

Reply:
The cost of upconversion is much more manageable. I have never seen a case where it was a problem. The key thing is that you need to decompress in any case in the produce side. For the fetch size, the difference between just sending the bytes versus decompressing, converting, recompressing is huge.

Reply:
There is some additional nuance here as well. If you upgrade producers before upgrading the message format, then the broker will have to downconvert before writing. That is just as costly as upconversion.

Reply:
Hmm... shouldn't broker negotiate the message format version with producer such that the atual message format version can not be greater than what broker configuration supports?

Reply:
That logic doesn't exist currently. Checking the message format requires using the config API. I agree it would be a nice improvement, but we don't have it today.